File: edge/src/lib/money.ts
// Decimal-safe money for the SalesFactory edge backend.
//
// Mirrors the Go libs/money: amounts are stored as an integer number of minor
// units (BigInt) at SCALE decimal places, never as JS number/float. Wire format
// is a decimal string ("600.66") + ISO-4217 currency code.

export const SCALE = 6;

const POW: bigint[] = Array.from({ length: SCALE + 1 }, (_, i) => 10n ** BigInt(i));

export class Money {
  readonly units: bigint;
  readonly currency: string;

  private constructor(units: bigint, currency: string) {
    this.units = units;
    this.currency = currency;
  }

  static zero(currency: string): Money {
    return new Money(0n, normCurrency(currency));
  }

  /** Parse a decimal string such as "600.66" into Money. Throws on invalid. */
  static parse(amount: string, currency: string): Money {
    const cur = normCurrency(currency);
    if (cur.length !== 3) throw new Error(`money: invalid currency ${JSON.stringify(currency)}`);
    let s = (amount ?? "").trim();
    if (s === "") throw new Error("money: empty amount");

    let neg = false;
    if (s[0] === "-") { neg = true; s = s.slice(1); }
    else if (s[0] === "+") s = s.slice(1);

    const dot = s.indexOf(".");
    let intPart = dot === -1 ? s : s.slice(0, dot);
    let fracPart = dot === -1 ? "" : s.slice(dot + 1);
    if (intPart === "" && fracPart === "") throw new Error(`money: invalid amount ${JSON.stringify(amount)}`);
    if (intPart === "") intPart = "0";
    if (!isDigits(intPart) || (fracPart !== "" && !isDigits(fracPart))) {
      throw new Error(`money: invalid amount ${JSON.stringify(amount)}`);
    }
    if (fracPart.length > SCALE) throw new Error(`money: amount ${JSON.stringify(amount)} exceeds ${SCALE} decimals`);

    const fracPadded = fracPart.padEnd(SCALE, "0");
    let units = BigInt(intPart + fracPadded);
    if (neg) units = -units;
    return new Money(units, cur);
  }

  isZero(): boolean { return this.units === 0n; }
  sign(): number { return this.units > 0n ? 1 : this.units < 0n ? -1 : 0; }

  add(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this.units + other.units, this.currency);
  }
  sub(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this.units - other.units, this.currency);
  }
  equals(other: Money): boolean {
    return this.currency === other.currency && this.units === other.units;
  }

  /** Format with `places` fractional digits, rounded half-up. */
  format(places = 2): string {
    if (places < 0) places = 0;
    if (places > SCALE) places = SCALE;
    const neg = this.units < 0n;
    let abs = this.units < 0n ? -this.units : this.units;

    if (places < SCALE) {
      const div = POW[SCALE - places];
      const half = div / 2n;
      const q = abs / div;
      const r = abs % div;
      abs = r >= half ? q + 1n : q;
    }
    let digits = abs.toString();
    if (places === 0) return (neg && abs !== 0n ? "-" : "") + digits;
    while (digits.length <= places) digits = "0" + digits;
    const intPart = digits.slice(0, digits.length - places);
    const fracPart = digits.slice(digits.length - places);
    return (neg && abs !== 0n ? "-" : "") + intPart + "." + fracPart;
  }

  toString(): string { return this.format(2); }

  private assertSameCurrency(other: Money): void {
    if (this.currency !== other.currency) {
      throw new Error(`money: currency mismatch ${this.currency} vs ${other.currency}`);
    }
  }
}

/** Multiply money by a decimal quantity string, rounding half-up to `decimals`. */
export function mulDecimal(m: Money, qty: string, decimals = 2): Money {
  return scaleRoundedToMoney(ratMul(m.units, parseRat(qty)), m.currency, decimals);
}

/** Compute m * (ratePercent/100), rounding half-up to `decimals`. */
export function percentOf(m: Money, ratePercent: string, decimals = 2): Money {
  const rate = parseRat(ratePercent);
  // (units * rateNum) / (rateDen * 100)
  return scaleRoundedToMoney(
    { num: m.units * rate.num, den: rate.den * 100n },
    m.currency,
    decimals,
  );
}

/** Divide money by a positive rate string (e.g. Bs -> USD at BCV), half-up. */
export function divByRate(m: Money, rate: string, targetCurrency: string, decimals = 2): Money {
  const r = parseRat(rate);
  if (r.num <= 0n) throw new Error(`money: invalid rate ${JSON.stringify(rate)}`);
  // m.units is at SCALE; result currency string built from rounded value.
  return scaleRoundedToMoney({ num: m.units * r.den, den: r.num }, normCurrency(targetCurrency), decimals);
}

// --- internal rational helpers (num/den of bigints) ---

type Rat = { num: bigint; den: bigint };

function parseRat(s: string): Rat {
  s = (s ?? "").trim();
  let neg = false;
  if (s[0] === "-") { neg = true; s = s.slice(1); }
  const dot = s.indexOf(".");
  if (dot === -1) {
    if (!isDigits(s || "0")) throw new Error(`invalid decimal ${JSON.stringify(s)}`);
    const n = BigInt(s || "0");
    return { num: neg ? -n : n, den: 1n };
  }
  const intPart = s.slice(0, dot) || "0";
  const fracPart = s.slice(dot + 1);
  if (!isDigits(intPart) || !isDigits(fracPart || "0")) throw new Error(`invalid decimal ${JSON.stringify(s)}`);
  const den = 10n ** BigInt(fracPart.length);
  const num = BigInt(intPart) * den + BigInt(fracPart || "0");
  return { num: neg ? -num : num, den };
}

// ratMul multiplies a money-units integer (at SCALE) by a rational, returning a
// rational still scaled at SCALE.
function ratMul(unitsAtScale: bigint, r: Rat): Rat {
  return { num: unitsAtScale * r.num, den: r.den };
}

// scaleRoundedToMoney takes a rational that represents a value already scaled by
// 10^SCALE (i.e. value*10^SCALE = num/den), rounds it to `decimals` places
// half-up, and returns Money.
function scaleRoundedToMoney(r: Rat, currency: string, decimals: number): Money {
  // value = (num/den) / 10^SCALE. We want round(value * 10^decimals).
  // = round( num / (den * 10^(SCALE-decimals)) )   [decimals <= SCALE]
  const shift = POW[SCALE - decimals];
  const denom = r.den * shift;
  const neg = (r.num < 0n) !== (denom < 0n);
  const absNum = r.num < 0n ? -r.num : r.num;
  const absDen = denom < 0n ? -denom : denom;
  let q = absNum / absDen;
  const rem = absNum % absDen;
  if (rem * 2n >= absDen) q += 1n;
  // q is the amount in 10^-decimals units; build a decimal string then parse.
  const dec = scaledIntToDecimal(neg ? -q : q, decimals);
  return Money.parse(dec, currency);
}

function scaledIntToDecimal(q: bigint, decimals: number): string {
  const neg = q < 0n;
  let digits = (q < 0n ? -q : q).toString();
  if (decimals === 0) return (neg ? "-" : "") + digits;
  while (digits.length <= decimals) digits = "0" + digits;
  const intPart = digits.slice(0, digits.length - decimals);
  const fracPart = digits.slice(digits.length - decimals);
  return (neg ? "-" : "") + intPart + "." + fracPart;
}

function normCurrency(c: string): string { return (c ?? "").trim().toUpperCase(); }
function isDigits(s: string): boolean { return s.length > 0 && /^[0-9]+$/.test(s); }

File: libs/money/money.go
// Package money provides decimal-safe monetary values for Meriplaza.
//
// Money is stored as an integer number of minor units (e.g. cents) plus an
// ISO-4217 currency code. This avoids binary floating-point error, which is a
// hard requirement for fiscal documents (see docs/venezuela-fiscal-invoice-api.md).
//
// API payloads carry amounts as decimal strings ("600.66"); this package parses
// and formats those strings without ever going through float64.
package money

import (
	"errors"
	"fmt"
	"math/big"
	"strings"
)

// Scale is the number of decimal places kept internally. Six places matches the
// invoice schema's money pattern (up to 6 fractional digits).
const Scale = 6

// Money is an immutable decimal amount in a single currency.
type Money struct {
	// units is the amount expressed in 10^-Scale minor units.
	units    *big.Int
	currency string
}

// ErrCurrencyMismatch is returned when two amounts in different currencies are
// combined.
var ErrCurrencyMismatch = errors.New("money: currency mismatch")

// Zero returns a zero amount in the given currency.
func Zero(currency string) Money {
	return Money{units: big.NewInt(0), currency: normCurrency(currency)}
}

// Parse converts a decimal string such as "600.66" into Money. The string may
// carry a leading minus sign and up to Scale fractional digits.
func Parse(amount, currency string) (Money, error) {
	cur := normCurrency(currency)
	if len(cur) != 3 {
		return Money{}, fmt.Errorf("money: invalid currency %q", currency)
	}
	s := strings.TrimSpace(amount)
	if s == "" {
		return Money{}, errors.New("money: empty amount")
	}

	neg := false
	switch s[0] {
	case '-':
		neg = true
		s = s[1:]
	case '+':
		s = s[1:]
	}

	intPart, fracPart, hasFrac := strings.Cut(s, ".")
	if intPart == "" && fracPart == "" {
		return Money{}, fmt.Errorf("money: invalid amount %q", amount)
	}
	if intPart == "" {
		intPart = "0"
	}
	if !isDigits(intPart) || (hasFrac && !isDigits(fracPart)) {
		return Money{}, fmt.Errorf("money: invalid amount %q", amount)
	}
	if len(fracPart) > Scale {
		return Money{}, fmt.Errorf("money: amount %q exceeds %d decimal places", amount, Scale)
	}

	// Right-pad the fractional part to Scale digits.
	fracPadded := fracPart + strings.Repeat("0", Scale-len(fracPart))

	units, ok := new(big.Int).SetString(intPart+fracPadded, 10)
	if !ok {
		return Money{}, fmt.Errorf("money: invalid amount %q", amount)
	}
	if neg {
		units.Neg(units)
	}
	return Money{units: units, currency: cur}, nil
}

// MustParse is like Parse but panics on error; use only for constants/tests.
func MustParse(amount, currency string) Money {
	m, err := Parse(amount, currency)
	if err != nil {
		panic(err)
	}
	return m
}

// Currency returns the ISO-4217 code.
func (m Money) Currency() string {
	if m.currency == "" {
		return ""
	}
	return m.currency
}

// IsZero reports whether the amount is exactly zero.
func (m Money) IsZero() bool { return m.units == nil || m.units.Sign() == 0 }

// Sign returns -1, 0, or +1.
func (m Money) Sign() int {
	if m.units == nil {
		return 0
	}
	return m.units.Sign()
}

// Add returns m+other. Currencies must match.
func (m Money) Add(other Money) (Money, error) {
	if err := m.assertSameCurrency(other); err != nil {
		return Money{}, err
	}
	return Money{units: new(big.Int).Add(m.intOrZero(), other.intOrZero()), currency: m.currency}, nil
}

// Sub returns m-other. Currencies must match.
func (m Money) Sub(other Money) (Money, error) {
	if err := m.assertSameCurrency(other); err != nil {
		return Money{}, err
	}
	return Money{units: new(big.Int).Sub(m.intOrZero(), other.intOrZero()), currency: m.currency}, nil
}

// Equal reports whether two amounts are identical in value and currency.
func (m Money) Equal(other Money) bool {
	return m.currency == other.currency && m.intOrZero().Cmp(other.intOrZero()) == 0
}

// Cmp compares amounts of the same currency, returning -1, 0, or +1.
func (m Money) Cmp(other Money) (int, error) {
	if err := m.assertSameCurrency(other); err != nil {
		return 0, err
	}
	return m.intOrZero().Cmp(other.intOrZero()), nil
}

// String formats the amount as a decimal string with no trailing-zero trimming
// beyond two places (e.g. "600.66", "0.00"). It always shows at least 2 places.
func (m Money) String() string {
	return m.Format(2)
}

// Format renders the amount with the given number of fractional digits (rounded
// half-up). Use 2 for display, Scale for full precision.
func (m Money) Format(places int) string {
	if places < 0 {
		places = 0
	}
	if places > Scale {
		places = Scale
	}
	u := m.intOrZero()
	neg := u.Sign() < 0
	abs := new(big.Int).Abs(u)

	// Round from Scale places down to the requested number of places.
	if places < Scale {
		div := pow10(Scale - places)
		half := new(big.Int).Div(div, big.NewInt(2))
		q, r := new(big.Int).DivMod(abs, div, new(big.Int))
		if r.Cmp(half) >= 0 {
			q.Add(q, big.NewInt(1))
		}
		abs = q
	}

	digits := abs.String()
	if places == 0 {
		s := digits
		if neg && abs.Sign() != 0 {
			s = "-" + s
		}
		return s
	}
	for len(digits) <= places {
		digits = "0" + digits
	}
	intPart := digits[:len(digits)-places]
	fracPart := digits[len(digits)-places:]
	s := intPart + "." + fracPart
	if neg && abs.Sign() != 0 {
		s = "-" + s
	}
	return s
}

func (m Money) assertSameCurrency(other Money) error {
	if m.currency != other.currency {
		return fmt.Errorf("%w: %q vs %q", ErrCurrencyMismatch, m.currency, other.currency)
	}
	return nil
}

func (m Money) intOrZero() *big.Int {
	if m.units == nil {
		return big.NewInt(0)
	}
	return m.units
}

func normCurrency(c string) string { return strings.ToUpper(strings.TrimSpace(c)) }

func isDigits(s string) bool {
	if s == "" {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

func pow10(n int) *big.Int {
	return new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(n)), nil)
}
